runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerPlugin.java

```diff
@@ -0,0 +1,413 @@
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+import javax.imageio.ImageIO;
+import javax.inject.Inject;
+import static net.runelite.api.AnimationID.FARMING_HARVEST_FLOWER;
+import static net.runelite.api.AnimationID.IDLE;
+import net.runelite.api.ChatMessageType;
+import net.runelite.api.Client;
+import net.runelite.api.GameState;
+import net.runelite.api.MenuAction;
+import net.runelite.api.ObjectComposition;
+import net.runelite.api.events.AnimationChanged;
+import net.runelite.api.events.ChatMessage;
+import net.runelite.api.events.GameStateChanged;
+import net.runelite.api.events.MenuOptionClicked;
+import net.runelite.client.config.ConfigManager;
+import net.runelite.client.game.ItemManager;
+import net.runelite.client.plugins.Plugin;
+import net.runelite.client.plugins.PluginDescriptor;
+import net.runelite.client.plugins.farmingtracker.data.FarmingTick;
+import net.runelite.client.plugins.farmingtracker.data.PatchLocation;
+import net.runelite.client.plugins.farmingtracker.data.PatchRowData;
+import net.runelite.client.plugins.farmingtracker.data.Seed;
+import net.runelite.client.plugins.farmingtracker.data.SeedStatus;
+import net.runelite.client.plugins.farmingtracker.ui.PatchList;
+import net.runelite.client.plugins.farmingtracker.ui.PatchRow;
+import net.runelite.client.task.Schedule;
+import net.runelite.client.ui.NavigationButton;
+import net.runelite.client.ui.PluginToolbar;
+import net.runelite.client.util.QueryRunner;
+import net.runelite.client.util.Text;
+
+@PluginDescriptor(
+	name = "Farming Tracker"
+)
+public class FarmingTrackerPlugin extends Plugin
+{
+	@Inject
+	private PluginToolbar pluginToolbar;
+
+	@Inject
+	private ItemManager itemManager;
+
+	@Inject
+	private Client client;
+
+	@Inject
+	private QueryRunner queryRunner;
+
+	@Inject
+	private ConfigManager configManager;
+
+	private final Set<String> firstOptions = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
+
+	private FarmingTrackerPanel panel;
+	private NavigationButton navButton;
+	private BufferedImage weeds;
+
+	private PatchLocation clickedOnPatch;
+	private ObjectComposition previousPatchObject;
+	private String menuTargetHardWood;
+	private int previousAnimation;
+	private int tickedAtEpoch = 0;
+
+	private final String CHAT_MESSAGE_PLANTED_SEED = "You plant ";
+	private final String CONFIG_GROUP = "farmingTracker";
+	private String lastUsername;
+
+	@Provides
+	FarmingTrackerConfig provideConfig(ConfigManager configManager)
+	{
+		return configManager.getConfig(FarmingTrackerConfig.class);
+	}
+
+	@Override
+	protected void startUp() throws Exception
+	{
+		firstOptions.addAll(Arrays.asList("pick", "harvest", "cure", "prune", "clear"));
+
+		BufferedImage icon;
+		synchronized (ImageIO.class)
+		{
+			icon = ImageIO.read(getClass().getResourceAsStream("farming.png"));
+			weeds = ImageIO.read(getClass().getResourceAsStream("weeds.png"));
+		}
+
+		panel = injector.getInstance(FarmingTrackerPanel.class);
```

Comment:
If you can, avoid creating spring components through Guice, it is always bad thing to do.